game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/change/CheckGeneralBattleEnd.java

```diff
@@ -0,0 +1,128 @@
+package games.strategy.triplea.delegate.battle.steps.change;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.DiceRoll;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.steps.BattleStep;
+import games.strategy.triplea.delegate.battle.steps.RetreatChecks;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class CheckGeneralBattleEnd implements BattleStep {
+  private static final long serialVersionUID = 5172121497955756220L;
+
+  private final BattleState battleState;
+
+  private final BattleActions battleActions;
+
+  protected BattleActions getBattleActions() {
+    return battleActions;
+  }
+
+  protected BattleState getBattleState() {
+    return battleState;
+  }
+
+  @Override
+  public List<String> getNames() {
+    return List.of();
+  }
+
+  @Override
+  public Order getOrder() {
+    return Order.GENERAL_BATTLE_END_CHECK;
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    if (hasSideLost(BattleState.Side.OFFENSE)) {
+      battleActions.endBattle(bridge);
+      battleActions.defenderWins(bridge);
+
+    } else if (hasSideLost(BattleState.Side.DEFENSE)) {
+      battleActions.endBattle(bridge);
+      battleActions.attackerWins(bridge);
+
+    } else if (isStalemate() && !canAttackerRetreatInStalemate()) {
+      battleActions.endBattle(bridge);
+      battleActions.nobodyWins(bridge);
+    }
+  }
+
+  protected boolean hasSideLost(final BattleState.Side side) {
+    return battleState.getUnits(side).stream().noneMatch(Matches.unitIsNotInfrastructure());
+  }
+
+  protected boolean isStalemate() {
+    return (battleState.getMaxBattleRounds() > 0
```

Comment:
I think we might be seeing a data clump code-smell here with getMaxBattleRounds & getBattleRounds().

Extract object is the refactoring pattern to fix that, in this case that would be a `BattleRound` object.

Then for example this code would be:
```
return battleState.getBattleRound().hasMoreRounds() || (getPower ...)
```

WDYT?

Reply:
`getBattleRound()` is already used and has an `int` return value.  So I used `getBattleRoundState()` unless you have a better name idea.

Reply:
I'm suggesting that `getBattleRound()` should return a `BattleRound` object. I''m not really suggesting to rename. The code smell is that we are always using both 'max rounds' and 'current rounds' together, that is the data clump (https://martinfowler.com/bliki/DataClump.html)

Reply:
I understand that.  But `getBattleRound()` already exists outside of the `BattleState` interface and is used by other classes that aren't battle steps.  So, to prevent having to change those classes, I modified `BattleState` to not expose `getBattleRound()` but to instead expose `getBattleRoundState()` and that returns the `BattleRound` object.

Reply:
Ah, understood. I wonder then if we actually have a true data clump if those other locations only need the current round.